drivers/adc/ad796x/ad796x.c

```diff
@@ -0,0 +1,297 @@
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the device gpios.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int ad796x_gpio_init(struct ad796x_dev *dev,
+			    struct ad796x_init_param *init_param)
+{
+	int ret, i;
+
+	struct no_os_gpio_init_param *params[] = {
+		init_param->gpio_adc_en3_fmc_ip,
+		init_param->gpio_adc_en2_fmc_ip,
+		init_param->gpio_adc_en1_fmc_ip,
+		init_param->gpio_adc_en0_fmc_ip,
+	};
+
+	struct no_os_gpio_desc *desc[] = {
+		dev->gpio_adc_en3_fmc,
+		dev->gpio_adc_en2_fmc,
+		dev->gpio_adc_en1_fmc,
+		dev->gpio_adc_en0_fmc,
+	};
+
+	uint8_t value[] = {
+		ad796x_modes[init_param->mode].en3,
+		ad796x_modes[init_param->mode].en2,
+		ad796x_modes[init_param->mode].en1,
+		ad796x_modes[init_param->mode].en0,
+	};
+
+	for (i = 0; i < 4; i++) {
+		ret = no_os_gpio_get_optional(&desc[i], params[i]);
+		if (ret)
+			goto error;
+
+		if (!desc[i])
+			continue;
+
+		ret = no_os_gpio_direction_output(desc[i], value[i]);
+		if (ret)
+			goto error;
+	}
+
+	return 0;
+error:
+	ad796x_gpio_remove(dev);
+	return ret;
+}
+
+/**
+ * @brief Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad796x_init(struct ad796x_dev **device,
+		struct ad796x_init_param *init_param)
+{
+	struct ad796x_dev *dev;
+	int ret;
+
+	dev = (struct ad796x_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	ret = ad796x_gpio_init(dev, init_param);
+	if (ret)
+		goto error_gpio;
+
+	ret = axi_clkgen_init(&dev->clkgen, init_param->clkgen_ip);
+	if (ret) {
+		pr_err("error: axi_clkgen_init error: %d\n", ret);
+		goto err_clkgen;
+	}
+
+	ret = axi_clkgen_set_rate(dev->clkgen, 125000000);
+	if (ret) {
+		pr_err("error: axi_clkgen_set_rate error: %d\n", ret);
+		goto err_set_rate;
+	}
+
+	ret = no_os_pwm_init(&dev->axi_pwm_0, init_param->axi_pwm_0_ip);
+	if (ret) {
+		pr_err("error: no_os_pwm_init error: %d\n", ret);
+		goto err_pwm_0;
+	}
+
+	ret = no_os_pwm_init(&dev->axi_pwm_1, init_param->axi_pwm_1_ip);
+	if (ret) {
+		pr_err("error: no_os_pwm_init error: %d\n", ret);
+		goto err_pwm_1;
+	}
+
+	ret = axi_adc_init(&dev->ad796x_core, init_param->ad796x_core_ip);
+	if (ret) {
+		pr_err("axi_adc_init error: %d\n", ret);
+		goto err_adc;
+	}
+
+	ret = axi_dmac_init(&dev->axi_dmac, init_param->dmac_ip);
+	if (ret) {
+		pr_err("axi_dmac_init error: %d\n", ret);
+		goto err_dmac;
+	}
+
+	*device = dev;
+
+	return 0;
+
+err_dmac:
+	axi_adc_remove(dev->ad796x_core);
+err_adc:
+	no_os_pwm_remove(dev->axi_pwm_1);
+err_pwm_1:
+	no_os_pwm_remove(dev->axi_pwm_0);
+err_pwm_0:
+err_set_rate:
+	axi_clkgen_remove(dev->clkgen);
+err_clkgen:
+	ad796x_gpio_remove(dev);
+error_gpio:
+	no_os_free(dev);
```

Comment:
I think it is safe to have a single error label (but please check before).
basically, if the init function has not been called, the desc will be null and the remove functions should check whether the desc is null or not. if it is null nothing will happen, if it is not null it will be freed.

Reply:
Ok, that just means that the convention is that  if the desc returns something that
something has to be freed (even if the int func fails). 

i guess i can add those checks in remove and just call remove then.

Reply:
Well... the convention is that if something is allocated and ret is not 0, it should be freed before returning.
However, seeing how we use calloc, all pointers are initialized with null, so if we call the remove function, even if memory has not been allocated, then the remove function should check for null and free the memory only if the pointer is not null.
Assuming that all remove functions check for null pointer before attempting to free it, it is safe to call the remove function even if the init has not been called. This will make the code easier to follow and shorter (a single label can be used).

Reply:
RE: if something is allocated and ret is not 0, it should be freed before returning

What i see from the no os-code is a bit confusing:

axi_dmac does not free before returning -1
(https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_dmac/axi_dmac.c#L349)

axi_adc_core frees the allocation before returning -1 But the the pointer is not set to null after the free. 
https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_adc_core/axi_adc_core.c#L677


So in case one of the int function call fails, im not sure we can trust the pointer to be freed, or
to be freed and null.

Reply:
regarding https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_dmac/axi_dmac.c#L349, it looks like the remove function should be called if ret is not 0. so the suggestion works great in this case. I do not know why this implementation was done like that. personally I would rework it into something like:
```
struct axi_dmac *dmac;

dmac = (struct axi_dmac *)no_os_calloc(1, sizeof(*dmac));
if (!dmac)
	return -1;

dmac->name = init->name;
dmac->base = init->base;
dmac->irq_option = init->irq_option;

int32_t status = axi_dmac_detect_caps(dmac);
if (status < 0)
	goto free;

*dmac_core = dmac;

return 0;

free:
	no_os_free(dmac);
	return -1;
```

regarding https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_adc_core/axi_adc_core.c#L677,
the descriptor remains null in case of an error, only in case of success the descriptor is set: https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_adc_core/axi_adc_core.c#L673

so basically, dev->ad796x_core will be null if init is not successful, and when calling the remove function nothing will happen. 
the problem I see is that axi_adc_remove does not check for null pointer, we should add that check I think.

Reply:
yes, ok. with those changes. 

BTW i have done it this way because the logic i was following was
to keep balanced number of init/remove. 

this would mean that:
- init will free its resources in case of fail, a failed init is not a init  (like you mention)
- remove should not be called if init was not called before
- remove should not be called for a failed init
- should not assume the state of the pointer

Reply:
I understand... this is not mandatory. in my case the code would look cleaner...
also, to me there is no real value in adding a duplicate label: `err_pwm_0`

Reply:
also, regarding my previous comment `the problem I see is that axi_adc_remove does not check for null pointer, we should add that check I think.` there is no problem with that API, it is safe to call free for null pointers.